iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.hpp

```diff
@@ -97,19 +97,33 @@ class ChunkDistributor
 
     /// @brief Deliver the provided shared chunk to all the stored chunk queues. The chunk will be added to the chunk
     /// history
-    /// @param[in] shared chunk to be delivered
+    /// @param[in] chunk is the SharedChunk to be delivered
     void deliverToAllStoredQueues(mepoo::SharedChunk chunk) noexcept;
 
     /// @brief Deliver the provided shared chunk to the provided chunk queue. The chunk will NOT be added to the chunk
     /// history
-    /// @param[in] chunk queue to which this chunk shall be delivered
-    /// @param[in] shared chunk to be delivered
+    /// @param[in] queue to which this chunk shall be delivered
+    /// @param[in] chunk is the SharedChunk to be delivered
     /// @return false if a queue overflow occured, otherwise true
+    /// @todo iox-#27 should this be made private? It's unsafe to call this outside of the ChunkDistributor since the
+    /// queue might not exist anymore when it is passed from outside to this method since the owning port might have
+    /// unsubscribed or there might be an ABA problem when the port which owns the queue was destroyed and newly created
     bool deliverToQueue(cxx::not_null<ChunkQueueData_t* const> queue, mepoo::SharedChunk chunk) noexcept;
 
+    /// @brief Deliver the provided shared chunk to the chunk queue with the provided ID. The chunk will NOT be added
+    /// to the chunk history
+    /// @param[in] uniqueQueueId is an unique ID which identifies the queue to which this chunk shall be delivered
+    /// @param[in] lastKnownQueueIndex is used for a fast lookup of the queue with uniqueQueueId; if the queue is not
+    /// found at the index, the queue is searched by iteration over all stored queues and the new index is stored in
+    /// this parameter
+    /// @param[in] chunk is the SharedChunk to be delivered
+    /// @return ChunkDistributorError if the queue was not found
+    cxx::expected<ChunkDistributorError>
+    deliverToQueue(const UniqueQueueId uniqueQueueId, uint32_t& lastKnownQueueIndex, mepoo::SharedChunk chunk) noexcept;
```

Comment:
@elfenpiff I did not yet implement this since I want some feedback before I do it and also write the tests. I'm not quite sure about this since the optimization with `lastKnownQueueIndex` must be a reference which actually points to a location in shared chunk. You can see that in `ClientPort`, which has this knowledge, but further down the abstraction it is unknown that the chunk is a `ResponseHeader`. One alternative would be to request the queue index for a unique queue id separately and store that one in the `ResponseHeader`. The `deliverToQueue` method would still need to do sanity checks on the index and potentially iterate over all stored queues but that would be rare and therefore have virtually zero performance impact. What do you think?

Reply:
@elfenpiff I think I go with the route to request the queue index before sending the response